### PR TITLE
Request decoupling

### PR DESCRIPTION
**Pull request by vangheem**
- [x] container
- [x] indexer
- [x] container_settings
- [x] db
- [x] participations/user auth
- [x] security policy
- [x] _db_write_enabled

### CODE REVIEW

**Comment by vangheem** [comment]
Ready for review.

File: docs/source/developer/security.md [hunk 1]
```diff
@@ -500,15 +500,15 @@ from guillotina.security.utils import settings_for_object
 settings_for_object(obj)
 
 # Code to get the Interaction object ( security object )
-from guillotina.interfaces import IInteraction
+from guillotina.interfaces import ISecurityPolicy
 
-interaction = IInteraction(request)
+policy = get_utility(ISecurityPolicy)
```
Comment (bloodbare) [inline docs/source/developer/security.md]: :)

File: guillotina/auth/utils.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+
+def set_authenticated_user(user):
+    policy = get_utility(ISecurityPolicy)
+    policy.invalidate_cache()
+    if user is not None:
+        if hasattr(user, 'roles') and 'guillotina.Authenticated' not in user.roles:
+            user.roles['guillotina.Authenticated'] = 1
+    task_vars.authenticated_user.set(user)
+    task_vars.authenticated_user_groups.set({})
```
Comment (bloodbare) [inline guillotina/auth/utils.py]: I think that groups should be a parameter of the user

File: guillotina/factory/content.py [hunk 3]
```diff
@@ -161,8 +160,6 @@ def storage(self):
         """
         create root object if necessary
         """
-        request = make_mocked_request('POST', '/')
-        request._db_write_enabled = True
         tm = self.get_transaction_manager()
         txn = await tm.begin()
```
Comment (bloodbare) [inline guillotina/factory/content.py]: Could we set explicit that is write ?

File: guillotina/json/serialize_content.py [hunk 4]
```diff
@@ -155,7 +155,7 @@ def check_permission(self, permission_name):
             if permission is None:
                 self.permission_cache[permission_name] = True
             else:
-                security = IInteraction(self.request)
+                security = get_utility(ISecurityPolicy)
```
Comment (bloodbare) [inline guillotina/json/serialize_content.py]: Security as utility is shared across records?

File: guillotina/json/serialize_content.py [hunk 5]
```diff
@@ -63,7 +63,7 @@ def __init__(self, context, request):
             behaviors.append(behavior_schema.__identifier__)
 
         result = {
-            '@id': IAbsoluteURL(self.context, self.request)(),
+            '@id': get_multi_adapter((self.context, self.request), IAbsoluteURL)(),
```
Comment (bloodbare) [inline guillotina/json/serialize_content.py]: Why?

File: guillotina/security/policy.py [hunk 6]
```diff
@@ -477,20 +418,15 @@ def _global_roles_for(self, principal):
     def _global_permissions_for(self, principal, permission):
         """On a principal (user + group) get global permissions."""
         groups = get_utility(IGroups)
-        if self.principal and principal == self.principal.id:
+        authenticated_princ = get_authenticated_user()
```
Comment (bloodbare) [inline guillotina/security/policy.py]: Could we avoid calling to context vars in this environment, maybe setting the principal on the init ?

File: guillotina/security/utils.py [hunk 7]
```diff
@@ -31,10 +30,8 @@ def get_roles_with_access_content(obj, request=None):
     """ Return the roles that has access to the content that are global roles"""
     if obj is None:
         return []
-    if request is None:
-        request = get_current_request()
-    interaction = IInteraction(request)
-    roles = interaction.cached_roles(obj, 'guillotina.AccessContent', 'o')
+    policy = get_utility(ISecurityPolicy)
```
Comment (bloodbare) [inline guillotina/security/utils.py]: Utility is shared across requests?

File: guillotina/task_vars.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+from guillotina.interfaces import IRegistry
+from guillotina.interfaces import IRequest
+
+
+request: ContextVar[Optional[IRequest]] = ContextVar('g_request', default=None)
+txn: ContextVar[Optional[ITransaction]] = ContextVar('g_txn', default=None)
+tm: ContextVar[Optional[ITransactionManager]] = ContextVar('g_tm', default=None)
+futures: ContextVar[Optional[Dict]] = ContextVar('g_futures', default=None)
+authenticated_user: ContextVar[Optional[IPrincipal]] = ContextVar('g_authenticated_user', default=None)
+authenticated_user_groups: ContextVar[Optional[IPrincipal]] = ContextVar(
```
Comment (bloodbare) [inline guillotina/task_vars.py]: Why?

File: guillotina/factory/content.py (hunk 3 above)
Comment (vangheem) [reply guillotina/factory/content.py]: Yup, you can specific on the transaction if it is `read_only` or not now.

File: guillotina/auth/utils.py (hunk 2 above)
Comment (vangheem) [reply guillotina/auth/utils.py]: This is now cache on principal instead.

File: docs/source/developer/security.md (hunk 1 above)
Comment (vangheem) [reply docs/source/developer/security.md]: changed to adapter

File: guillotina/json/serialize_content.py (hunk 4 above)
Comment (vangheem) [reply guillotina/json/serialize_content.py]: switched to adapter.

File: guillotina/security/policy.py (hunk 6 above)
Comment (vangheem) [reply guillotina/security/policy.py]: done.

File: guillotina/task_vars.py (hunk 8 above)
Comment (vangheem) [reply guillotina/task_vars.py]: `authenticated_user_groups` removed.

File: guillotina/json/serialize_content.py (hunk 5 above)
Comment (vangheem) [reply guillotina/json/serialize_content.py]: using get_object_url now

**Pull request merged**